pools/usecase/pools_usecase.go

```diff
@@ -540,7 +546,78 @@ func (p *poolsUseCase) CalcExitCFMMPool(poolID uint64, exitingShares osmomath.In
 
 	// fine to pass empty context as no data is mutated
 	exitFee := pool.GetExitFee(sdk.Context{})
-	return pool.CalcExitPoolCoinsFromShares(sdk.Context{}, exitingShares, exitFee)
+
+	return calcExitPool(sdk.Context{}, pool, exitingSharesIn, exitFee)
+}
+
+// errMsgFormatSharesLargerThanMax is the error message format for when the exiting shares are larger than the max allowed.
+const errMsgFormatSharesLargerThanMax = "cannot exit all shares in a pool. Attempted to exit %f shares, max allowed is %f"
+
+// calcExitPool is a helper function to calculate the exit pool.
+// It is a direct port of the CalcExitPool function from the Gamm module with some modifications
+// to optimize the calculation for performance.
+func calcExitPool(ctx sdk.Context, pool types.CFMMPoolI, exitingSharesIn osmomath.Int, exitFee osmomath.Dec) (sdk.Coins, error) {
+	totalShares, err := strconv.ParseFloat(pool.GetTotalShares().String(), 64)
+	if err != nil {
+		return sdk.Coins{}, err
+	}
+
+	exitingShares, err := strconv.ParseFloat(exitingSharesIn.String(), 64)
+	if err != nil {
+		return sdk.Coins{}, err
+	}
+
+	if exitingShares >= totalShares {
+		return sdk.Coins{}, errorsmod.Wrapf(types.ErrLimitMaxAmount, errMsgFormatSharesLargerThanMax, exitingShares, totalShares-float64(osmomath.OneInt().Int64()))
+	}
+
+	// refundedShares = exitingShares * (1 - exit fee)
+	// with 0 exit fee optimization
+	var refundedShares float64
+	if !exitFee.IsZero() {
+		f, err := exitFee.Float64()
+		if err != nil {
+			return sdk.Coins{}, err
+		}
+
+		// exitingShares * (1 - exit fee)
+		oneSubExitFee := 1.0 - f
+		refundedShares = exitingShares * oneSubExitFee
+	} else {
+		refundedShares = exitingShares
+	}
+
+	shareOutRatio := refundedShares / totalShares
+
+	// exitedCoins = shareOutRatio * pool liquidity
+	poolLiquidity := pool.GetTotalPoolLiquidity(ctx)
+	exitedCoins := make(sdk.Coins, 0, len(poolLiquidity))
+
+	for _, asset := range poolLiquidity {
+		// round down here, due to not wanting to over-exit
+		amount := float64(asset.Amount.BigInt().Int64())
```

Comment:
I think this might overflow and panic - we should consider using string conversation like you've done in another example

Reply:
Might be worth adding a test to cover the overflow

Reply:
Good catch! Updated and included owerflow test. :+1: